src/services/safe_border.py

```diff
@@ -0,0 +1,164 @@
+        earliest_slashed_epoch_predicted = self._predict_earliest_slashed_epoch(validators_with_earliest_exit_epoch[0])
+
+        if earliest_slashed_epoch_predicted:
+            return earliest_slashed_epoch_predicted
+        return self._find_earliest_slashed_epoch(validators_with_earliest_exit_epoch, blockstamp)
+
+    # If there is no so many exited validators in line we can be quite sure that
+    # slashing has started not earlier than 8,192 epochs or ~36 days ago
+    def _predict_earliest_slashed_epoch(self, validator):
+        exit_epoch = int(validator.validator.validator.exit_epoch)
+        withdrawable_epoch = int(validator.validator.validator.withdrawable_epoch)
+
+        exited_period = withdrawable_epoch - exit_epoch
+        is_slashed_epoch_undetectable = exited_period > MIN_VALIDATOR_WITHDRAWABILITY_DELAY
+        if is_slashed_epoch_undetectable:
+            return None
+
+        return withdrawable_epoch - EPOCHS_PER_SLASHINGS_VECTOR
+
+    def _find_earliest_slashed_epoch(self, validators, blockstamp: BlockStamp):
+        pubkeys = get_validators_pubkeys(validators)
+        withdrawable_epoch = min(get_validators_withdrawable_epochs(validators))
+
+        last_finalized_request_id_slot = self._get_last_finalized_withdrawal_request_slot(blockstamp)
+
+        start_slot = max(last_finalized_request_id_slot, self._get_validators_earliest_activation_epoch(validators))
+        end_slot = min(blockstamp.ref_slot, self.get_epoch_first_slot(withdrawable_epoch - EPOCHS_PER_SLASHINGS_VECTOR))
+
+        while start_slot + 1 < end_slot:
+            mid_slot = (end_slot + start_slot) // 2
+            validators = self._get_archive_lido_validators_by_keys(mid_slot, pubkeys)
+            slashed_validators = filter_slashed_validators(validators)
+
+            if len(slashed_validators) > 0:
+                end_slot = mid_slot - 1
+            else:
+                start_slot = mid_slot + 1
+
+        return self.get_epoch_by_slot(end_slot)
+
+    def _filter_validators_with_earliest_exit_epoch(self, validators):
+        if len(validators) == 0:
+            return []
+
+        sorted_validators = sorted(validators, key = lambda validator: (int(validator.validator.validator.exit_epoch)))
+        earliest_exit_epoch = sorted_validators[0].validator.validator.exit_epoch
+        return filter_validators_by_exit_epoch(sorted_validators, earliest_exit_epoch)
+
+    def _get_validators_earliest_activation_epoch(self, validators):
+        if len(validators) == 0:
+            return 0
+
+        sorted_validators = sorted(validators, key = lambda validator: (int(validator.validator.validator.activation_epoch)))
+        return sorted_validators[0].validator.validator.activation_epoch
+
+    def _get_lido_validators(self, blockstamp: BlockStamp) -> list[LidoValidator]:
+        return self.w3.cc.get_validators(blockstamp.ref_slot)
```

Comment:
get_validators - Returns Validator type, not LidoValidator.
Use self.w3.cc.lido_validators.get_lido_validators